### PR TITLE
refactor(view): introduce AppViewManager to consolidate logic

### PR DESCRIPTION
**Pull request by tbosch**
AppViewManager is the single entry point to changing the view hierarchy.
It is split between the manager itself which does coordination and
helper methods, so both are easily testable in isolation.

Part of #1477 

After this is done the same pattern should be applied to the render side as well!

### CODE REVIEW

File: modules/angular2/src/core/compiler/dynamic_component_loader.js [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
+import {Key, Injector, Injectable, ResolvedBinding, Binding, bind} from 'angular2/di'
 import {Compiler} from './compiler';
 import {DirectiveMetadataReader} from './directive_metadata_reader';
 import {Type, BaseException, stringify, isPresent} from 'angular2/src/facade/lang';
 import {Promise} from 'angular2/src/facade/async';
 import {Component} from 'angular2/src/core/annotations/annotations';
-import {ViewFactory} from 'angular2/src/core/compiler/view_factory';
-import {AppViewHydrator} from 'angular2/src/core/compiler/view_hydrator';
-import {ElementRef, DirectiveBinding} from './element_injector';
+import {AppViewManager} from 'angular2/src/core/compiler/view_manager';
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/dynamic_component_loader.js]: General comment (no action required), our use of 'App' is in consistent, here it is meant as an opposite to 'render', but here: 
https://github.com/angular/angular/blob/ada1e642c57f19ad5cecde57db1aca1727bf2325/modules/angular2/src/core/application.js#L58
it is meant as a 'top-level'.

File: modules/angular2/src/core/compiler/view_container.js [hunk 2]
```diff
@@ -6,103 +6,68 @@ import {isPresent, isBlank} from 'angular2/src/facade/lang';
...
-
-    // The order in this list matches the DOM order.
-    this._views = [];
+  constructor(viewManager: avmModule.AppViewManager,
+              parentView: viewModule.AppView,
+              boundElementIndex: number,
+              defaultProtoView: viewModule.AppProtoView) {
+    this._viewManager = viewManager;
+    this._parentView = parentView;
+    this._boundElementIndex = boundElementIndex;
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_container.js]: It is odd that `ViewContainer` doesn't have a ref to `InternalViewContainer` but instead has to go through an index and a list in parentView. Any reason for this decision?

File: modules/angular2/src/core/compiler/view_manager.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  _utils:AppViewManagerUtils;
+  _renderer:Renderer;
+
+  constructor(viewPool:AppViewPool, utils:AppViewManagerUtils, renderer:Renderer) {
+    this._renderer = renderer;
+    this._viewPool = viewPool;
+    this._utils = utils;
+  }
+
+  createDynamicComponentView(hostView:viewModule.AppView, boundElementIndex:number,
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_manager.js]: Why not use `ElementRef` instead of `boundElementIndex`?

File: modules/angular2/src/core/compiler/view_manager.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  _createView(protoView:viewModule.AppProtoView) {
+    var view = this._viewPool.getView(protoView);
+    if (isBlank(view)) {
+      view = this._utils.createView(protoView, this);
+      var binders = protoView.elementBinders;
+      for (var binderIdx = 0; binderIdx < binders.length; binderIdx++) {
+        var binder = binders[binderIdx];
+        if (binder.hasStaticComponent()) {
+          var childView = this._createView(binder.nestedProtoView);
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_manager.js]: Shouldn't this be '_createViewRecurse' since it recurses here?

File: modules/angular2/src/core/compiler/view_manager_utils.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+      parentView.changeDetector.removeChild(hostView.changeDetector);
+      ListWrapper.remove(parentView.imperativeHostViews, hostView);
+    }
+  }
+
+  attachViewInContainer(parentView:viewModule.AppView, boundElementIndex:number, atIndex:number, view:viewModule.AppView) {
+    parentView.changeDetector.addChild(view.changeDetector);
+    var viewContainer = parentView.viewContainers[boundElementIndex];
+    if (isBlank(viewContainer)) {
+      viewContainer = new viewModule.InternalAppViewContainer();
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_manager_utils.js]: Why are we building it on demand. Shouldn't all containers be known in advance as injectable parameters to decorators, etc?

File: modules/angular2/src/core/compiler/view_manager_utils.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+      shadowDomAppInjector = appInjector;
+    }
+    return shadowDomAppInjector;
+  }
+
+  _hydrateView(view:viewModule.AppView, appInjector:Injector, hostElementInjector:eli.ElementInjector, context: Object, parentLocals:Locals) {
+    if (isBlank(appInjector)) {
+      appInjector = hostElementInjector.getShadowDomAppInjector();
+    }
+    if (isBlank(appInjector)) {
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_manager_utils.js]: why is this line necessary? Can we make `hostElementInjector.getShadowDomAppInjector()` always return non-null?

File: modules/angular2/src/core/compiler/view_manager_utils.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+          view.locals.set(exportImplicitName, elementInjector.getNgElement().domElement);
+        }
+      }
+    }
+
+  }
+
+  _setUpEventEmitters(view:viewModule.AppView, elementInjector:eli.ElementInjector, boundElementIndex:number) {
+    var emitters = elementInjector.getEventEmitterAccessors();
+    for(var directiveIndex = 0; directiveIndex < emitters.length; ++directiveIndex) {
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view_manager_utils.js]: `for (var`

File: modules/angular2/src/test_lib/test_bed.js [hunk 8]
```diff
@@ -112,12 +103,12 @@ export class TestBed {
  * Proxy to `AppView` return by `createView` in {@link TestBed} which offers a high level API for tests.
  */
 export class ViewProxy {
```
Comment (rkirov) [inline modules/angular2/src/test_lib/test_bed.js]: Not in this PR, but can we replace this test specific `ViewProxy` with a general `AppViewRef`. The apis used for tests should be similar to what we provide to users.

File: modules/angular2/src/core/compiler/view_container.js (hunk 2 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_container.js]: We create the `InternalViewContainer` lazily when the first `AppView` is inserted into it. Before this, we don't need it...

File: modules/angular2/src/core/compiler/view_manager.js (hunk 3 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_manager.js]: `ElementRef` contains a lot more than we need here...

File: modules/angular2/src/core/compiler/view_manager.js (hunk 4 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_manager.js]: Good point.

File: modules/angular2/src/core/compiler/view_manager_utils.js (hunk 5 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_manager_utils.js]: They can be asked for via `ElementRef.viewContainer` for an element that does not inject a `ViewContainer` itself (needed for overlay service).

File: modules/angular2/src/test_lib/test_lib.dart [hunk 9]
```diff
@@ -87,7 +91,32 @@ class Expect extends gns.Expect {
   void toImplement(expected) => toBeA(expected);
   void toBeNaN() => _expect(double.NAN.compareTo(actual) == 0, equals(true));
   void toHaveText(expected) => _expect(elementText(actual), expected);
+  void toHaveBeenCalledWith([a = _u, b = _u, c = _u, d = _u, e = _u, f = _u]) =>
+      _expect(_firstArgsMatch(actual, a, b, c, d, e, f), true, reason: 'method invoked with correct arguments');
   Function get _expect => gns.guinness.matchers.expect;
+
+  // TODO(vsavkin): move this hack into Guinness
```
Comment (rkirov) [inline modules/angular2/src/test_lib/test_lib.dart]: @vsavkin can you take a look at this.

File: modules/angular2/test/core/compiler/dynamic_component_loader_spec.js [hunk 10]
```diff
@@ -24,7 +24,7 @@ import {If} from 'angular2/src/directives/if';
 import {DirectDomRenderer} from 'angular2/src/render/dom/direct_dom_renderer';
 
 export function main() {
-  describe('DynamicComponentLoader', function () {
+  ddescribe('DynamicComponentLoader', function () {
```
Comment (rkirov) [inline modules/angular2/test/core/compiler/dynamic_component_loader_spec.js]: remove `ddescribe`

File: modules/angular2/test/core/compiler/integration_spec.js [hunk 11]
```diff
@@ -38,7 +38,7 @@ import {ElementRef} from 'angular2/src/core/compiler/element_injector';
 import {DirectDomRenderer} from 'angular2/src/render/dom/direct_dom_renderer';
 
 export function main() {
-  describe('integration tests', function() {
+  ddescribe('integration tests', function() {
```
Comment (rkirov) [inline modules/angular2/test/core/compiler/integration_spec.js]: same

File: modules/angular2/src/core/compiler/view_manager_utils.js (hunk 6 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_manager_utils.js]: This is for letting a call overwrite the default injector (i.e. prevent the use of `hostElementInjector.getShadowDomAppInjector()`). Needed e.g. for creating a `ProtoView` in a `ViewContainer` with a pre filled injector (e.g. the router does this).

File: modules/angular2/test/core/compiler/view_manager_spec.js [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+      binder.nestedProtoView = nestedProtoView;
+      return binder;
+    }
+
+    function createProtoView(binders = null) {
+      if (isBlank(binders)) {
+        binders = [];
+      }
+      var staticChildComponentCount = 0;
+      for (var i=0; i<binders.length; i++) {
```
Comment (rkirov) [inline modules/angular2/test/core/compiler/view_manager_spec.js]: syntax for iteration should be `for (var i = 0; i < binders.length; i++)`
